converting scripts/prepare_int_data.sh to its python version

keli78 opened the pull request:
convert scripts/prepare_int_data.sh to its python version

danpovey committed:
Various code-level changes, not yet tested, to support count pruning.

danpovey committed:
Various fixes.  Reduce number of files used in self_test.

danpovey committed:
minor fixes

danpovey committed:
some bug-fixes regarding enforcing min-counts and other recent changes

danpovey committed:
Various bug-fixes, and add a script in egs/self_test/local/run_mincount.sh to test the derivatives with min-counts.

danpovey committed:
Various minor script fixes

Commit:
converting scripts/prepare_int_data.sh to its python version

danpovey committed:
Add capability to enforce mincounts, to get_counts.py.  Various related bug-fixes.

danpovey committed:
A couple minor, cosmetic script fixes.

danpovey commented on scripts/prepare_int_data.py:
```diff
@@ -0,0 +1,185 @@
+  sys.exit("prepare_int_data.py: expected text_to_int.py to be on the path")
+
+# create the output data directory
+if not os.path.exists(args.int_dir):
```
you can remove this if-block... it's redundant with the next line. [makedirs makes it recursively.]

danpovey commented on scripts/prepare_int_data.py:
```diff
@@ -0,0 +1,185 @@
+  return output
+
+def GetNames(script_dir, text_dir, int_dir):
+  command = "{0}/internal/get_names.py {1} > {2}/names".format(
```
I think it would be more natural to handle the path issue by adding script_dir/internal to the path up at the top of the program [os.environ['PATH']...] rather than using the absolute path.

danpovey commented on scripts/prepare_int_data.py:
```diff
@@ -0,0 +1,185 @@
+  RunCommand(command)
+
+def GetNumTrainSets(int_dir):
+  command = "cat {0}/names | wc -l".format(int_dir)
```
I don't think using 'cat' and 'ls'  is a very Pythonic way to do this, would be better to iterate over lines using readline() or readlines().
But regardless, it would be easier to get this number from the file 'num_train_sets', which will exist in the input directory.
See how other scripts read num_train_sets.

danpovey commented on scripts/prepare_int_data.py:
```diff
@@ -0,0 +1,185 @@
+dest = args.int_dir + "/words.txt"
+CopyFile(src, dest)
+
+# return the number of train sets 
```
this is not a function, nothing is returning.

danpovey commented on scripts/prepare_int_data.py:
```diff
@@ -0,0 +1,185 @@
+with open(args.int_dir + "/num_train_sets", "w") as file:
+  file.write(str(GetNumTrainSets(args.int_dir)))
+
+# return num-words in vocab 
```
again, nothing is returning.
And here and above, it's good practice to close your file handles.

danpovey commented on scripts/prepare_int_data.py:
```diff
@@ -0,0 +1,185 @@
+with open(args.int_dir + "/num_words", "w") as file:
+  file.write(str(GetNumWords(args.vocab)))
+
+src = args.int_dir + "/names"
```
What you are doing here is a very unnatural way to accomplish this.  The shell script was iterating over a pair of (int, name) pairs, and adding the pair ('dev', 'dev').  It would be much more natural to do this manually as python code.
In addition, using the name 'dest' in such a wide scope is very confusing; having more specific variable names would have made things much clearer (but most of this code should be rewritten).

danpovey commented on scripts/prepare_int_data.py:
```diff
@@ -0,0 +1,185 @@
+def PrepareData(line):
+  [int, name] = line.split()
+  if os.path.exists(args.text_dir + "/" + name + ".txt.gz"):
+    command = "set -o pipefail; gunzip -c {text_dir}/{name}.txt.gz | text_to_int.py {vocab} | gzip -c > {int_dir}/{int}.txt.gz 2>{int_dir}/log/{int}.log".format(
```
please break this long line, e.g. by using backslash `\` or parentheses.
Likewise else there.  Don't go too far past 80 characters.

keli78 replied:
Just modified the code based on your comments.
Ke

On Sat, Jun 25, 2016 at 6:52 PM, Daniel Povey notifications@github.com
wrote:

> In scripts/prepare_int_data.py
> https://github.com/danpovey/pocolm/pull/18#discussion_r68495029:
> 
> > -  line = subprocess.check_output(command, shell = True)
> > -  try:
> > -    a = line.split()
> > -    assert len(a) == 2
> > -    ans = int(a[1])
> > -  except:
> > -    sys.exit("prepare_int_data.py: error: unexpected output '{0}' from command {1}".format(
> > -                line, command))
> > -  return ans
> >   +
> >   +# we can include the preparation of the dev data in the following
> >   +# by adding "dev dev" to the contents of int_dir/names
> >   +def PrepareData(line):
> > -  [int, name] = line.split()
> > -  if os.path.exists(args.text_dir + "/" + name + ".txt.gz"):
> > -    command = "set -o pipefail; gunzip -c {text_dir}/{name}.txt.gz | text_to_int.py {vocab} | gzip -c > {int_dir}/{int}.txt.gz 2>{int_dir}/log/{int}.log".format(
> 
> please break this long line, e.g. by using backslash \ or parentheses.
> Likewise else there. Don't go too far past 80 characters.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/danpovey/pocolm/pull/18/files/b4e574c7a96866d2347764cc620ecaaddd049c5c#r68495029,
> or mute the thread
> https://github.com/notifications/unsubscribe/ANVxSkHbGPyNTr9qN1aw3T3OxM-yE-ofks5qPbEYgaJpZM4I-PTe
> .

## 

Ke Li
Dept. of Electrical and Computer Engineering
Johns Hopkins University
Email: kli26@jhu.edu

Commit:
modified scripts/prepare_int_data.py

danpovey committed:
Some script reorganizations for improved logging; changes to how counts are computed, to make use of multiple cores (for speed).

danpovey commented on scripts/prepare_int_data.py:
```diff
@@ -0,0 +1,146 @@
+  if os.system(command) != 0:
+    ExitProgram("prepare_int_data.py: error running command: " + command)
+
+def GetCommandStdout(command):
```
Today I have changed how these things are done: these functions like GetCommandStdout should be imported from internal/pocolm_common.py, and you should use a log/ subdirectory in the destination dir.  Please do 'git pull' and see how it's done.  I'll check after that.

keli78 replied:
Hi Dan,
I have changed the python code after doing "git pull". Since I use "with
open() as f", I don't need to close the file by "f.close()".
Ke

On Sun, Jun 26, 2016 at 8:19 PM, Daniel Povey notifications@github.com
wrote:

> In scripts/prepare_int_data.py
> https://github.com/danpovey/pocolm/pull/18#discussion_r68513136:
> 
> > +for name in [ 'num_train_sets', 'num_words' ]:
> > -    f = open(args.int_dir + os.sep + name)
> > -    globals()[name] = int(f.readline())
> > -    f.close()
> >   +
> >   +def ExitProgram(message):
> > -  print(message, file=sys.stderr)
> > -  os._exit(1)
> >   +
> >   +def RunCommand(command):
> > -  # print the command for logging
> > -  print(command, file=sys.stderr)
> > -  if os.system(command) != 0:
> > -    ExitProgram("prepare_int_data.py: error running command: " + command)
> >   +
> >   +def GetCommandStdout(command):
> 
> Today I have changed how these things are done: these functions like
> GetCommandStdout should be imported from internal/pocolm_common.py, and you
> should use a log/ subdirectory in the destination dir. Please do 'git pull'
> and see how it's done. I'll check after that.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/danpovey/pocolm/pull/18/files/93b41726ac53e0157eee398d637ea5a4ffbb89e4#r68513136,
> or mute the thread
> https://github.com/notifications/unsubscribe/ANVxSjblI7AP4dE1947DpjnH95GqzBHbks5qPxcqgaJpZM4I-PTe
> .

## 

Ke Li
Dept. of Electrical and Computer Engineering
Johns Hopkins University
Email: kli26@jhu.edu

Commit:
Merge branch 'master' of https://github.com/danpovey/pocolm into prepare-int-data

Commit:
modified scripts/prepare_int_data.py after /internal/pocolm_common.py is ready

Commit:
modified scripts/prepare_int_data.py after /internal/pocolm_common.py is ready

danpovey commented:
OK.
So I assume you've checked that at least some of the scripts still work as before?
I don't like having merges from master into feature branches; can you please rebase, as follows:

```
git checkout prepare-int-data  # assuming that's what you called your branch
git rebase master  # assuming you have a branch called master
```

and then push with --force to your feature branch on github.

keli78 commented:
Yes, I replaced the python script with the bash script on swbd and it works
fine ( I got the same perplexities using prepare_int_data.sh). Do you want
me to check other recipes before rebasing?
Ke

On Mon, Jun 27, 2016 at 3:14 PM, Daniel Povey notifications@github.com
wrote:

> OK.
> So I assume you've checked that at least some of the scripts still work as
> before?
> I don't like having merges from master into feature branches; can you
> please rebase, as follows:
> 
> git checkout prepare-int-data  # assuming that's what you called your branch
> git rebase master  # assuming you have a branch called master
> 
> and then push with --force to your feature branch on github.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/danpovey/pocolm/pull/18#issuecomment-228845845, or mute
> the thread
> https://github.com/notifications/unsubscribe/ANVxSmVzMr1PsR16yNS4QA2sesNRqd7Qks5qQCECgaJpZM4I-PTe
> .

## 

Ke Li
Dept. of Electrical and Computer Engineering
Johns Hopkins University
Email: kli26@jhu.edu

danpovey commented:
Try at least the self_test, because that one uses multiple data sources; it
might exercise more code.

On Mon, Jun 27, 2016 at 12:19 PM, Ke Li notifications@github.com wrote:

> Yes, I replaced the python script with the bash script on swbd and it works
> fine ( I got the same perplexities using prepare_int_data.sh). Do you want
> me to check other recipes before rebasing?
> Ke
> 
> On Mon, Jun 27, 2016 at 3:14 PM, Daniel Povey notifications@github.com
> wrote:
> 
> > OK.
> > So I assume you've checked that at least some of the scripts still work
> > as
> > before?
> > I don't like having merges from master into feature branches; can you
> > please rebase, as follows:
> > 
> > git checkout prepare-int-data # assuming that's what you called your
> > branch
> > git rebase master # assuming you have a branch called master
> > 
> > and then push with --force to your feature branch on github.
> > 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > https://github.com/danpovey/pocolm/pull/18#issuecomment-228845845, or
> > mute
> > the thread
> > <
> > https://github.com/notifications/unsubscribe/ANVxSmVzMr1PsR16yNS4QA2sesNRqd7Qks5qQCECgaJpZM4I-PTe
> > 
> > .
> 
> ## 
> 
> Ke Li
> Dept. of Electrical and Computer Engineering
> Johns Hopkins University
> Email: kli26@jhu.edu
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/danpovey/pocolm/pull/18#issuecomment-228847323, or mute
> the thread
> https://github.com/notifications/unsubscribe/ADJVu9xohfEhgfR8QT2Sa6g54ldjuvowks5qQCJVgaJpZM4I-PTe
> .

keli78 commented:
Just fixed two bugs after testing on self_test.
I have a quick question. When I compare the two output directories using
"diff -qr dir1 dir2", it gives the results as below:

keli1@t01:/export/a16/keli1/pocolm-shpy/egs/self_test$ diff -qr
data/500_3/int data/500_3/int_new

Files data/500_3/int/1.txt.gz and data/500_3/int_new/1.txt.gz differ

Files data/500_3/int/2.txt.gz and data/500_3/int_new/2.txt.gz differ

Files data/500_3/int/3.txt.gz and data/500_3/int_new/3.txt.gz differ

Files data/500_3/int/4.txt.gz and data/500_3/int_new/4.txt.gz differ

Files data/500_3/int/5.txt.gz and data/500_3/int_new/5.txt.gz differ

Files data/500_3/int/6.txt.gz and data/500_3/int_new/6.txt.gz differ

Files data/500_3/int/7.txt.gz and data/500_3/int_new/7.txt.gz differ

Files data/500_3/int/8.txt.gz and data/500_3/int_new/8.txt.gz differ

Files data/500_3/int/9.txt.gz and data/500_3/int_new/9.txt.gz differ

Files data/500_3/int/dev.txt.gz and data/500_3/int_new/dev.txt.gz differ

Then I use "zdiff" and it turns out that they are the same. Not sure why
they're different by using "diff".

Ke

On Mon, Jun 27, 2016 at 3:26 PM, Daniel Povey notifications@github.com
wrote:

> Try at least the self_test, because that one uses multiple data sources; it
> might exercise more code.
> 
> On Mon, Jun 27, 2016 at 12:19 PM, Ke Li notifications@github.com wrote:
> 
> > Yes, I replaced the python script with the bash script on swbd and it
> > works
> > fine ( I got the same perplexities using prepare_int_data.sh). Do you
> > want
> > me to check other recipes before rebasing?
> > Ke
> > 
> > On Mon, Jun 27, 2016 at 3:14 PM, Daniel Povey notifications@github.com
> > wrote:
> > 
> > > OK.
> > > So I assume you've checked that at least some of the scripts still work
> > > as
> > > before?
> > > I don't like having merges from master into feature branches; can you
> > > please rebase, as follows:
> > > 
> > > git checkout prepare-int-data # assuming that's what you called your
> > > branch
> > > git rebase master # assuming you have a branch called master
> > > 
> > > and then push with --force to your feature branch on github.
> > > 
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > https://github.com/danpovey/pocolm/pull/18#issuecomment-228845845,
> > > or
> > > mute
> > > the thread
> > > <
> > 
> > https://github.com/notifications/unsubscribe/ANVxSmVzMr1PsR16yNS4QA2sesNRqd7Qks5qQCECgaJpZM4I-PTe
> > 
> > > .
> > 
> > ## 
> > 
> > Ke Li
> > Dept. of Electrical and Computer Engineering
> > Johns Hopkins University
> > Email: kli26@jhu.edu
> > 
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > https://github.com/danpovey/pocolm/pull/18#issuecomment-228847323, or
> > mute
> > the thread
> > <
> > https://github.com/notifications/unsubscribe/ADJVu9xohfEhgfR8QT2Sa6g54ldjuvowks5qQCJVgaJpZM4I-PTe
> > 
> > .
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/danpovey/pocolm/pull/18#issuecomment-228848898, or mute
> the thread
> https://github.com/notifications/unsubscribe/ANVxSpDy-gqUBKmmg64uSiVNjP7q06Q9ks5qQCPUgaJpZM4I-PTe
> .

## 

Ke Li
Dept. of Electrical and Computer Engineering
Johns Hopkins University
Email: kli26@jhu.edu

danpovey commented:
It's OK; AFAIK gzip does not guarantee to compress in a deterministic way
[or maybe you ran it from a different machine].  Or maybe they really were
differerent because the code  changed in the interim.
Dan

On Mon, Jun 27, 2016 at 1:16 PM, Ke Li notifications@github.com wrote:

> Just fixed two bugs after testing on self_test.
> I have a quick question. When I compare the two output directories using
> "diff -qr dir1 dir2", it gives the results as below:
> 
> keli1@t01:/export/a16/keli1/pocolm-shpy/egs/self_test$ diff -qr
> data/500_3/int data/500_3/int_new
> 
> Files data/500_3/int/1.txt.gz and data/500_3/int_new/1.txt.gz differ
> 
> Files data/500_3/int/2.txt.gz and data/500_3/int_new/2.txt.gz differ
> 
> Files data/500_3/int/3.txt.gz and data/500_3/int_new/3.txt.gz differ
> 
> Files data/500_3/int/4.txt.gz and data/500_3/int_new/4.txt.gz differ
> 
> Files data/500_3/int/5.txt.gz and data/500_3/int_new/5.txt.gz differ
> 
> Files data/500_3/int/6.txt.gz and data/500_3/int_new/6.txt.gz differ
> 
> Files data/500_3/int/7.txt.gz and data/500_3/int_new/7.txt.gz differ
> 
> Files data/500_3/int/8.txt.gz and data/500_3/int_new/8.txt.gz differ
> 
> Files data/500_3/int/9.txt.gz and data/500_3/int_new/9.txt.gz differ
> 
> Files data/500_3/int/dev.txt.gz and data/500_3/int_new/dev.txt.gz differ
> 
> Then I use "zdiff" and it turns out that they are the same. Not sure why
> they're different by using "diff".
> 
> Ke
> 
> On Mon, Jun 27, 2016 at 3:26 PM, Daniel Povey notifications@github.com
> 
> wrote:
> 
> > Try at least the self_test, because that one uses multiple data sources;
> > it
> > might exercise more code.
> > 
> > On Mon, Jun 27, 2016 at 12:19 PM, Ke Li notifications@github.com
> > wrote:
> > 
> > > Yes, I replaced the python script with the bash script on swbd and it
> > > works
> > > fine ( I got the same perplexities using prepare_int_data.sh). Do you
> > > want
> > > me to check other recipes before rebasing?
> > > Ke
> > > 
> > > On Mon, Jun 27, 2016 at 3:14 PM, Daniel Povey <
> > > notifications@github.com>
> > > wrote:
> > > 
> > > > OK.
> > > > So I assume you've checked that at least some of the scripts still
> > > > work
> > > > as
> > > > before?
> > > > I don't like having merges from master into feature branches; can you
> > > > please rebase, as follows:
> > > > 
> > > > git checkout prepare-int-data # assuming that's what you called your
> > > > branch
> > > > git rebase master # assuming you have a branch called master
> > > > 
> > > > and then push with --force to your feature branch on github.
> > > > 
> > > > —
> > > > You are receiving this because you authored the thread.
> > > > Reply to this email directly, view it on GitHub
> > > > https://github.com/danpovey/pocolm/pull/18#issuecomment-228845845,
> > > > or
> > > > mute
> > > > the thread
> > > > <
> > 
> > https://github.com/notifications/unsubscribe/ANVxSmVzMr1PsR16yNS4QA2sesNRqd7Qks5qQCECgaJpZM4I-PTe
> > 
> > > > .
> > > 
> > > ## 
> > > 
> > > Ke Li
> > > Dept. of Electrical and Computer Engineering
> > > Johns Hopkins University
> > > Email: kli26@jhu.edu
> > > 
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > https://github.com/danpovey/pocolm/pull/18#issuecomment-228847323,
> > > or
> > > mute
> > > the thread
> > > <
> > 
> > https://github.com/notifications/unsubscribe/ADJVu9xohfEhgfR8QT2Sa6g54ldjuvowks5qQCJVgaJpZM4I-PTe
> > 
> > > .
> > 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > https://github.com/danpovey/pocolm/pull/18#issuecomment-228848898, or
> > mute
> > the thread
> > <
> > https://github.com/notifications/unsubscribe/ANVxSpDy-gqUBKmmg64uSiVNjP7q06Q9ks5qQCPUgaJpZM4I-PTe
> > 
> > .
> 
> ## 
> 
> Ke Li
> Dept. of Electrical and Computer Engineering
> Johns Hopkins University
> Email: kli26@jhu.edu
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/danpovey/pocolm/pull/18#issuecomment-228862161, or mute
> the thread
> https://github.com/notifications/unsubscribe/ADJVu7kCoFnDBFFjj5RI4KeGS3JaXuM0ks5qQC-ugaJpZM4I-PTe
> .

Commit:
fixed two file iteration bugs

Commit:
fixed two file iteration bugs

danpovey commented:
The rebase did not work.  It looks like you now even have conflicts that you have not properly resolved.
The commands I gave made some assumptions, including that your 'master' branch was up to date.  If you ran them with out-of-date master, you might have got something a bit like this.
Maybe do some reading about git
https://git-scm.com/book/en/v2/Getting-Started-Git-Basics
